brainframe/cli/commands/self_update.py

```diff
@@ -0,0 +1,113 @@
+import os
+import stat
+import sys
+from argparse import ArgumentParser
+from pathlib import Path
+from typing import Optional, Tuple, Union
+
+import i18n
+import requests
+from brainframe.cli import __version__, config, frozen_utils, print_utils
+from packaging import version
+
+from .utils import command
+
+_RELEASES_URL_PREFIX = "https://{subdomain}aotu.ai"
+_BINARY_URL = "{prefix}/releases/brainframe-cli/brainframe"
+_LATEST_TAG_URL = "{prefix}/releases/brainframe-cli/latest"
+
+
+@command("self-update")
+def self_update():
+    _parse_args()
+
+    if not frozen_utils.is_frozen():
+        print_utils.fail_translate("self-update.not-frozen")
+
+    executable_path = Path(sys.executable)
+
+    # Check if the user has permissions to overwrite the executable in its
+    # current location
+    if not os.access(executable_path, os.W_OK):
+        error_message = i18n.t(
+            "general.file-bad-write-permissions", path=executable_path
+        )
+        error_message += "\n"
+        error_message += i18n.t("general.retry-as-root")
+        print_utils.fail(error_message)
+
+    credentials = config.staging_credentials()
+
+    prefix = _RELEASES_URL_PREFIX.format(
+        subdomain="staging." if config.is_staging.value else ""
+    )
+    binary_url = _BINARY_URL.format(prefix=prefix)
+
+    current_version = version.parse(__version__)
+    latest_version = _latest_version(prefix, credentials)
+
+    if current_version >= latest_version:
+        print_utils.fail_translate(
+            "self-update.already-up-to-date",
+            current_version=current_version,
+            latest_version=latest_version,
+        )
+
+    # Get the updated executable
+    print_utils.translate("self-update.downloading")
+    response = requests.get(binary_url, auth=credentials, stream=True)
+    if not response.ok:
+        print_utils.fail_translate(
+            "self-update.error-downloading",
+            status_code=response.status_code,
+            error_message=response.text,
+        )
+
+    # Overwrite the existing executable with the new one
+    os.remove(executable_path)
```

Comment:
Would it be hard to download into a TemporaryDirectory _then_ remove the executable? That way i the internet is crappy and the request fails while downloading it's nbd

Reply:
Good idea!